_drafts/git-bisect.md

```diff
@@ -0,0 +1,611 @@
+<span style="color:red;">f21a4c55 (tag v1.2) Add comments and static-ize flasher.c</span>
+<span style="color:red;">47a89108 Actually improved performance</span>
+<span style="color:red;">0afd538f Reverting performance improvements that made things slower</span>
+<span style="color:red;">84c98ac9 Performance improvements</span>
+<span style="color:red;">26338040 Refactor python script with classes</span>
+<span style="color:red;">54526f51 (tag v1.1) Note the work that still needs to be done in the readme</span>
+<span style="color:red;">9f0d9a91 Move flash driver to non-memory-mapped</span>
+<span style="color:red;">e717c312 Repo linting changes</span>
+<span style="color:red;">a2b8a742 More misc. changes</span>
+<span style="color:red;">e0c258c0 Misc. Changes</span>
+<span style="color:yellow;">cecc054a Integrate all the things</span>
+<span style="color:green;">acf78303 (tag v1.0) Add initialization procedure</span>
+</code>
+</pre>
+</div>
+</div>
+
+Now, at long last, let’s test our last commit and tell git.
+
+```
+>>> python update.py
+Flashing successful!
+>>> git bisect good
+e0c258c05343202c53a294ba59704bb02565c9d5 is the first bad commit
+```
+
+Cool, we found the bad commit, and if we look at our world one last time, we
+see exactly where things went bad.
+
+<div class='highlighter-rouge'>
+<div class='highlight'>
+<pre>
+<code>
+<span style="color:red;">e1952a1c Comment flasher files</span>
+<span style="color:red;">51ed84b6 Comment main.c</span>
+<span style="color:red;">b3a10580 Helpful python error message and traceback</span>
+<span style="color:red;">f21a4c55 (tag v1.2) Add comments and static-ize flasher.c</span>
+<span style="color:red;">47a89108 Actually improved performance</span>
+<span style="color:red;">0afd538f Reverting performance improvements that made things slower</span>
+<span style="color:red;">84c98ac9 Performance improvements</span>
+<span style="color:red;">26338040 Refactor python script with classes</span>
+<span style="color:red;">54526f51 (tag v1.1) Note the work that still needs to be done in the readme</span>
+<span style="color:red;">9f0d9a91 Move flash driver to non-memory-mapped</span>
+<span style="color:red;">e717c312 Repo linting changes</span>
+<span style="color:red;">a2b8a742 More misc. changes</span>
+<span style="color:green;">e0c258c0 Misc. Changes</span>
+<span style="color:green;">cecc054a Integrate all the things</span>
+<span style="color:green;">acf78303 (tag v1.0) Add initialization procedure</span>
+</code>
+</pre>
+</div>
+</div>
+
+### Fixing the Bug
+As we saw at the end of the last section, our bug seems to have been introduced
+with commit e0c258c0. Let’s check that changeset for verification changes...
+
+```diff
++ uint8_t readback_data[size];
++ nor_flash_dev->operations.read(sector_to_compare, readback_data, size);
++ int return_code = memcmp((void *) DATA_PAYLOAD_ADDR, readback_data, size);
++ return (return_code == 0)
+```
+
```

Comment:
Maybe: Might + in the function signature as part of the diff so its easier to follow along with your description below. I think that would be something like:

```diff
+// @return 0 if contents match
+ int verify_copy(uint32_t sector_to_compare) {
[...]
+     uint8_t readback_data[size];
+     nor_flash_dev->operations.read(sector_to_compare, readback_data, size);
+     int return_code = memcmp((void *) DATA_PAYLOAD_ADDR, readback_data, size);
+     return (return_code == 0)
+}
```

Reply:
Good call, will do.

Reply:
Done!